### PR TITLE
Fix Concat operations with variables #953

### PR DESCRIPTION
**Pull request by Slamdunk**
Fix NOT https://github.com/sebastianbergmann/php-code-coverage/issues/953

### CODE REVIEW

File: tests/_files/source_with_multiline_constant_return.php [hunk 1]
```diff
@@ -396,7 +396,7 @@ public function unaryLogicalNotWithNotConstInTheMiddle(): string
     public function unaryMinusWithNotConstInTheMiddle(): string
     {
         return -
-        (
+        (int) (
```
Comment (mvorisek) [inline tests/_files/source_with_multiline_constant_return.php]: cast is tested in a following method, this method should test "unary minus with not const in the middle"

File: tests/_files/source_with_multiline_constant_return.php [hunk 2]
```diff
@@ -485,10 +485,30 @@ public function unaryMinusNowdoc(): bool
     {
         return
             -
+            (int)
```
Comment (mvorisek) [inline tests/_files/source_with_multiline_constant_return.php:488]: no cast needed here too, and the type can be changed to `int`

File: tests/_files/source_with_multiline_constant_return.php [hunk 3]
```diff
@@ -485,10 +485,30 @@ public function unaryMinusNowdoc(): bool
...
+        $var1 =
+            'right'
+            .
+            $var1
+        ;
+
+        $var1 =
+            $var1
+            .
+            'left'
```
Comment (mvorisek) [inline tests/_files/source_with_multiline_constant_return.php]: instead of string use nowdoc here, I expect a wrong line then

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php [hunk 4]
```diff
@@ -165,6 +165,14 @@ private function getLines(NodeAbstract $node, bool $fromReturns): array
         }
 
         if ($node instanceof BinaryOp) {
+            if ($node instanceof BinaryOp\Concat &&
```
Comment (mvorisek) [inline src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: how is concat special vs. other binary operators line plus?

File: tests/_files/source_with_multiline_constant_return.php (hunk 1 above)
Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php]: We need valid source code for proper debugging, and this doesn't alter the validity of the test

File: tests/_files/source_with_multiline_constant_return.php (hunk 3 above)
Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php]: We should use both then

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php (hunk 4 above)
Comment (Slamdunk) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: I have no idea, but it seems it is special.

I guess it is because it can trigger typecasting on object by a `__toString()` call, wether other binary ops don't

File: tests/_files/source_with_multiline_constant_return.php (hunk 3 above)
Comment (mvorisek) [reply tests/_files/source_with_multiline_constant_return.php]: I would say nowdoc is enough as we test "AST processing" is used instead of "node stard/end line"

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php (hunk 4 above)
Comment (mvorisek) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: Your reasoning might make sense, but I still belive the var can be optimized to constant with opcache and thus var. itself should not generate a coverage line. I need to test.

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php [hunk 5]
```diff
@@ -165,6 +165,14 @@ private function getLines(NodeAbstract $node, bool $fromReturns): array
...
+                (
+                    $node->left instanceof Node\Expr\Variable ||
+                $node->right instanceof Node\Expr\Variable
+                )) {
+                return [$node->right->getStartLine()];
```
Comment (mvorisek) [inline src/StaticAnalysis/ExecutableLinesFindingVisitor.php]:
```suggestion
                return [$this->getNodeStartLine($node->right)];
```

will fix start line for everything, string, nowdoc, ... never use fixed start/end line on a generic node

File: tests/_files/source_with_multiline_constant_return.php (hunk 1 above)
Comment (mvorisek) [reply tests/_files/source_with_multiline_constant_return.php]: it alters - unary minus of concat (const expr), not unary minus of cast expr

File: tests/_files/source_with_multiline_constant_return.php [hunk 6]
```diff
@@ -485,10 +485,30 @@ public function unaryMinusNowdoc(): bool
...
             -
+            (int)
             <<<'EOF'
                 foo
                 foo
                 EOF
         ;
     }
+
+    public function concatWithVar(): string
```
Comment (mvorisek) [inline tests/_files/source_with_multiline_constant_return.php:496]: binary plus operator needs same test

File: tests/_files/source_with_multiline_constant_return.php (hunk 1 above)
Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php]: It makes no sense to test something that doesn't even compile.

File: tests/_files/source_with_multiline_constant_return.php (hunk 3 above)
Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php]: :+1:

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php (hunk 5 above)
Comment (Slamdunk) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php]: Done :+1:

File: tests/_files/source_with_multiline_constant_return.php (hunk 1 above)
Comment (mvorisek) [reply tests/_files/source_with_multiline_constant_return.php]: it compiles fine: https://3v4l.org/aaq3l

Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php]: Indeed this work, but the other doesn't

File: tests/_files/source_with_multiline_constant_return.php (hunk 2 above)
Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php:488]: `Uncaught TypeError: Unsupported operand types: string * int`

Comment (Slamdunk) [reply tests/_files/source_with_multiline_constant_return.php:488]: Without the typecast, it doesn't compile

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php (hunk 4 above)
Comment (mvorisek) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: I had to revert it, as it is unstable (line is not present in xdebug) when the var expr is known (from constatnt expr).

File: tests/_files/source_with_multiline_constant_return.php (hunk 6 above)
Comment (mvorisek) [reply tests/_files/source_with_multiline_constant_return.php:496]: reverted, see https://github.com/mvorisek/php-code-coverage/pull/2#discussion_r1019558371

File: src/StaticAnalysis/ExecutableLinesFindingVisitor.php (hunk 4 above)
Comment (Slamdunk) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: I have to say, I'm kind of disappointed on how you are handling this contribution.

I created a valid test for a valid use case, I got the test passing, you found another edge case and istead of adding them to the tests, you throw my contribution away?

I'm losing my willingness to help, if you discard my help

Comment (mvorisek) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: I am sorry to hear that, I already spent about 2 - 3 MDs on static analyser to be really a minimal subset of xdebug and implementing https://github.com/sebastianbergmann/php-code-coverage/issues/953#issuecomment-1310831008 might take another several MDs. So what I was supposed to do?

> instead of adding them to the tests

test is added in https://github.com/sebastianbergmann/php-code-coverage/blob/c4630fdec1def41d63ce7844d6e91ced5ccfa478/tests/_files/source_with_heavy_indentation.php#L95-L110 (`$xa` is non-const assign, `$xb` is const, not an uncommon case)

Comment (Slamdunk) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]:
```php
$var1 =
    'right'
    .
    $var1
;

$var1 =
    $var1
    .
    'left'
;
```
Where exactly are these two cases tested now in https://github.com/sebastianbergmann/php-code-coverage/pull/949 ?

Comment (mvorisek) [reply src/StaticAnalysis/ExecutableLinesFindingVisitor.php:169]: (as of PHP 8.0) the variable assignments are covered

the concat const expr does not produce a coverage line if the variable is const expr - this is tested a for ex. in l105 in the example I posted - I also tested locally concat is not different than any other binary operator (like plus) in sense of coverage
